src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -123,9 +125,33 @@ protected override DiagnosticTag[] ConvertTags(DiagnosticData diagnosticData)
                 return ImmutableArray<IDiagnosticSource>.Empty;
             }
 
+            if (nonLocalDocumentDiagnostics)
+                return GetNonLocalDiagnosticSources();
+
             return taskList
                 ? ImmutableArray.Create<IDiagnosticSource>(new TaskListDiagnosticSource(document, globalOptions))
                 : ImmutableArray.Create<IDiagnosticSource>(new DocumentDiagnosticSource(diagnosticKind, document));
+
+            ImmutableArray<IDiagnosticSource> GetNonLocalDiagnosticSources()
+            {
+                Debug.Assert(!taskList);
+
+                // Compiler does not report any non-local diagnostics, so we bail out when those are requested.
+                // Additionally, all analyzer reported non-local diagnostics are classified as semantic diagnostics,
+                // so we bail out when only analyzer syntax diagnostics are requested.
+                if (diagnosticKind is DiagnosticKind.CompilerSyntax or DiagnosticKind.CompilerSemantic or DiagnosticKind.AnalyzerSyntax)
```

Comment:
> Note that currently vscode does not specify diagnostic kinds so it will always be DiagnosticKind.All. We could add support for it though in similar way that we're adding support for this source.

I did not that when I added the unit test. I kept these as more of defensive checks, but now that I think more it's probably best to just have an assert that `diagnosticKind` is `DiagnosticKind.All` on this code path.

> I'm wondering if it might make it cleaner to have an additional diagnostic kind category, instead of non-local being its own separate thing

That will add confusion to existing usages of `DiagnosticKind` enum in our code base. For example, the existing 4 values in this enum matches with the 4 diagnostic kind queries made in VS LSP pull diagnostics world for document diagnostics.